### PR TITLE
Replace laravel auth app with mock-oauth2-server running in docker

### PR DESCRIPTION
**Pull request by patcon**
Closes: https://github.com/GCTC-NTGC/gc-digital-talent/issues/2755

Supercedes: https://github.com/GCTC-NTGC/gc-digital-talent/issues/2716 and https://github.com/GCTC-NTGC/gc-digital-talent/pull/2741

This PR removes our old auth app and replaces it with a mock oauth2 server. Best way to review is to step through commits, as the removal commit is quite a bit noisier than the rest :)

## Screencast

[Link to full video](https://recordit.co/ybRVpzESLS)

## Steps to review
1. checkout branch
2. copy new lines from `api/.env.example` into `api/.env` (override prior local auth values)
4. run `docker-compose up --detach --build` to spin up the new `mock-auth` container (and rebuild `php` container)
5. ~~add this line to your hosts file: `127.0.0.1 auth.talent.canada.test`~~
6. run `docker-compose run --rm maintenance bash refresh_api.sh` to install new composer dependencies

## To Dos
- [x] add ability for `api` to work without `x5c` jwks prop
- [x] added and configured mock-oauth2-server docker container
- [x] removed all code and references to `auth` app
- [x] confirm `bcmath` is on UAT or can be -- @petertgiles  
    - check `php -m` (and find `bcmath`)
    - ✅ confirmed `gmp` and `bcmath` loaded for CLI php on development server
    - ~~if not there, and no package for something like `php-bcmath`, then we can easily switch to `gmp`, which may be simpler: https://stackoverflow.com/questions/40010197/how-to-install-gmp-for-php7-on-ubuntu~~ 
- [x] more documentation?
- peter feedback
    - [x] spin out new issue for consolidating jwt lib https://github.com/GCTC-NTGC/gc-digital-talent/issues/2835
    - [x] remove dup web-token/jwt-core entry from composer.json https://github.com/GCTC-NTGC/gc-digital-talent/pull/2813#discussion_r877401389
    - [x] proxy auth server via apache https://github.com/GCTC-NTGC/gc-digital-talent/pull/2813#discussion_r877404786
    - [x] hardcode client_id https://github.com/GCTC-NTGC/gc-digital-talent/pull/2813#discussion_r877408664
    - [x] remove final bits of code for laravel auth https://github.com/GCTC-NTGC/gc-digital-talent/pull/2813#discussion_r877463611
    - [x] update `maintenance/README.md` https://github.com/GCTC-NTGC/gc-digital-talent/pull/2813#discussion_r877469902
    - [ ] document how mock auth server works
        - https://github.com/GCTC-NTGC/gc-digital-talent/pull/2813#discussion_r878633347
        - https://github.com/GCTC-NTGC/gc-digital-talent/pull/2813#discussion_r878677779
    - [x] re-ticket issue for writing e2e tests https://github.com/GCTC-NTGC/gc-digital-talent/issues/2671
    - [ ] push this feature branch to dev -- @petertgiles @patcon

### CODE REVIEW

**Comment by petertgiles** [comment]
Nice, looking forward to trying this out!  It uses bcmath?  Any confirmation on whether we can put this on our production server?

**Comment by esizer** [comment]
Just dropping in to say this is cool! 😎  Reviewed to try and learn a bit more about how we are handling auth 👍

**Comment by patcon** [comment]
@tristan-orourke requested review from tristan, since there's UAT server steps, but anyone else is welcome to try before we merge. cc @petertgiles

**Comment by patcon** [comment]
> Oh wait, I guess the mock auth won't have to run on a production server.

@petertgiles no no, you're right, because the `bcmath` ext is something we need for the cryptographic operations done by `api` app. This allows us to avoid requiring the arcane `x5c` prop (giving a full cert). Even google's own jwks endpoint doesn't offer a full cert in `x5c`, and instead builds the public key from `n` and `e`: https://www.googleapis.com/oauth2/v3/certs

File: api/app/Services/OpenIdBearerTokenService.php
Comment (petertgiles) [inline api/app/Services/OpenIdBearerTokenService.php:93]: There's some other Laravel Auth stuff left over here that can be removed:
- `api/app/Services/LocalAuthBearerTokenService.php`
- `api/tests/Unit/LocalAuthBearerTokenTest.php`

Additionally, this file can have the branching logic removed: `api/app/Providers/BearerTokenServiceProvider.php` which would allow you to get rid of `api/app/Services/Contracts/BearerTokenServiceInterface.php` entirely.

File: infrastructure/conf/mock-auth-server.json
Comment (petertgiles) [inline infrastructure/conf/mock-auth-server.json:7]: Can this mock issue access tokens and refresh tokens?  Out of scope for here I think but it would be worth putting on the board if possible since we definitely use that with SiC.

File: maintenance/README.md
Comment (petertgiles) [inline maintenance/README.md:16]: Please update the instructions above for step 3.

File: infrastructure/php-container/Dockerfile
Comment (petertgiles) [inline infrastructure/php-container/Dockerfile:18]: I'm so worried this will be a showstopper.  😟

**Review by petertgiles — CHANGES_REQUESTED** [review]
Whoah, I love this so much.  :heart: :heart: :heart:. Another patcon wizard PR.  I love seeing ` +321 −31,458` at the top of the PR page.  I've got a few "questions, concerns, threats" below but I'm so excited for this.

**Comment by patcon** [comment]
This is a killer review peter! Will get on this tomorrow :)

File: api/app/Services/OpenIdBearerTokenService.php
Comment (patcon) [reply api/app/Services/OpenIdBearerTokenService.php:12]: Done! https://github.com/GCTC-NTGC/gc-digital-talent/issues/2835

File: api/composer.json
Comment (patcon) [reply api/composer.json:21]: Oof. Some sort of merge issue for sure. Seems the lockfile resolution occurred with the [appropriate] 2.2 version, so just going to delete from composer.json

File: api/composer.json
Comment (patcon) [inline api/composer.json:13]:
```suggestion
```

File: api/composer.json
Comment (patcon) [reply api/composer.json]: Done!

File: docker-compose.yml
Comment (patcon) [reply docker-compose.yml:62]: There is [some mention of proxy support in a merged PR](https://github.com/navikt/mock-oauth2-server/pull/187#issue-1112784096), though this [hostname usage is recommended for docker in the README](https://github.com/navikt/mock-oauth2-server#docker-compose)

I'd messed around with this, but it didn't reveal itself to be trivial at the time.

But when I tried just with curl, I had some better luck, just by adding the header `X-Forwarded-Proto: http`

```
patcon at Patricks-MacBook-Pro in ~/repos/gc-digital-talent (mock-oauth2-server-alt●)
$ curl -s http://localhost:8081/default/.well-known/openid-configuration --header 'Host: localhost:8080' | grep issuer
  "issuer" : "http://auth.talent.canada.test:8081/default",

patcon at Patricks-MacBook-Pro in ~/repos/gc-digital-talent (mock-oauth2-server-alt●)
$ curl -s http://localhost:8081/default/.well-known/openid-configuration --header 'Host: localhost:8080' --header 'X-Forwarded-Proto: http' | grep issuer
  "issuer" : "http://localhost:8080/default",
```

I'll take a run at this again :)

File: docker-compose.yml
Comment (patcon) [reply docker-compose.yml:70]: Ah, I like that -- I guess I didn't want ppl to have "bad" experiences with the new thing, but maybe it's better not to hide this stuff, as it'll make us less familiar with the way production works :)

And... when developers ask for bad experiences, GIVE THEM BAD EXPERIENCES haha

But I spent a bit fidgeting with it, and seems that we can only match based on a single field, so no locking the redirect_url. Managing/validating secrets and locking to specific redirect_uri doensn't seem to be something this mock server can do.

This mock servers vibe is that it's stateless and kinda just "agrees with everything asked" (ie. anything that would require state) and checks cryptographic validations. So it's assuming that the client is managing expirations and things, and it's not double-checking on any of that-- just giving the client new tokens when asked, and doing all the oauth dance.

It mostly just offers a properly signed JWT with a specific "sub" (aka email) that our system believes has certain privileges :)

I think we can fake more complex responses with `cy.intercept()` if we want to validate how our app acts in those situations. This mock is most for operating like everything is dandy :)

Is this helpful context? If so, I can add a step to document it :)

File: infrastructure/conf/mock-auth-server.json
Comment (patcon) [reply infrastructure/conf/mock-auth-server.json:7]: yep, i believe that part of the process is already happening automatically behind the scenes, at least it seems to be happening from the auth server logs :)

as i half-ass rambled above, my understanding is this: whenever the access token expires, it fetches a new access token from the auth server using the refresh token, all according to the oauth2 dance

File: infrastructure/php-container/Dockerfile
Comment (patcon) [reply infrastructure/php-container/Dockerfile:18]: I'm hopeful between the two we'll get it working 😬  (also, we can always use another pure-php library to rebuild the public key from JWKS data -- this just makes it easier)

File: maintenance/README.md
Comment (patcon) [reply maintenance/README.md:18]: Thanks! Done.

File: api/app/Services/OpenIdBearerTokenService.php
Comment (patcon) [reply api/app/Services/OpenIdBearerTokenService.php:93]: I might need a little guidance with pulling out the interface. Might you be able to offer a short diff I removing it in one place? (assuming that's easier than typing out an explanation or hopping on a quick call, both of which I'd also be grateful for!)

**Comment by patcon** [comment]
K, I think I address all the feedback, or am waiting for a response on that one part 🙌

File: api/app/Services/OpenIdBearerTokenService.php
Comment (petertgiles) [reply api/app/Services/OpenIdBearerTokenService.php:93]: Yeah, good chance to do some pairing. :wink: Let's look at it today sometime.

**Comment by patcon** [comment]
@petertgiles Ok, figured out how to get mock-oauth2-server working behind apache proxy. It's a little more than I expected, and I need to work outside htaccess to get proxy working, but lemme know if it looks ok!

The main confusing bit (for which I tried to document in case anyone wants to unwind what's happening later), is that both the API code (running inside the container) and the host browser (running outside the container) need to be able to find the auth server at the same url. It's explained best in comments within `infrastructure/php-container/mock-auth-proxy.apache-site.conf`, so i won't repeat myself here :)

**Comment by patcon** [comment]
My attempt to document our auth flow a little bit (WIP):

```mermaid
sequenceDiagram
    autonumber
    participant browser as Browser (User)
    participant admin as React Admin App
    participant api as Laravel API App
    participant auth as Auth Server

    Note right of auth: localStorage:{stored_locale: "fr"}
    Note right of auth: cookies: none
    browser->>admin: GET /admin
    admin-->>browser: 200 OK. App HTML & JS.

    Note left of browser: check localStorage.stored_locale
    browser->>browser: /fr/admin
    Note left of browser: check localStorage.id_token (null)
    Note left of browser: render Login button

    browser->>api: GET /login?from=/fr/admin&locale=fr
    api-->>browser: 302 Found. Set-Cookie: laravel_session=<...>, Location: <...>
    Note right of auth: cookies: laravel_session=<...>
    Note left of browser: Redirect

    browser->>auth: GET /oxauth/authorize? ... {client_id,redirect_uri,response_type,scope,state,nonce,acr_values,ui_locales}
    auth-->>browser: 200 OK. Login Form.

    Note left of browser: Form field: username=admin@test.com
    browser->>auth: POST /oxauth/authorize? ... {client_id,redirect_uri,response_type,scope,state,nonce,acr_values,ui_locales} username=admin@test.com
    auth-->>browser: 302 Redirect. Location: <...>
    Note left of browser: Redirect

    browser->>api: GET /auth-callback?code=...&state=...
    Note right of api: check cookie: laravel_session
    api-->>browser: 302 Redirect. Location: <...>
    Note left of browser: Redirect

    browser->>admin: GET /fr/admin? ... {token_type,id_token, access_token, refresh_token,expires_in}
    admin-->>browser: 200 OK.
    Note right of auth: localStorage:{stored_locale: "fr", id_token: "<...>"}

    browser->>browser: /fr/admin
    Note left of browser: check localStorage.id_token (present)
    Note left of browser: render Logout button

    browser->>api: POST /graphql
    Note left of browser: TODO token validation.
    api-->>browser: 200 OK.
```

**Comment by patcon** [comment]
aside from my stretch goal of writing some high level docs in response to some of peter's questions, this is ready for review! (i also stubbed out some ideas in #2671 for end-to-end tests the would exercise some security edge-cases and ensure we handle those

File: docker-compose.yml
Comment (patcon) [reply docker-compose.yml]: Fixed!

**Comment by petertgiles** [comment]
I'm having some trouble with this.  I'm getting a 503 on the authorize request.  Any idea what the trouble is?  I rebuilt my images and containers and reran setup.sh.
image
Here are my Apache logs:
```
[Thu May 26 14:19:30.136440 2022] [proxy:error] [pid 19] (111)Connection refused: AH00957: http: attempt to connect to 172.19.0.2:8000 (mock-auth) failed
[Thu May 26 14:19:30.136464 2022] [proxy_http:error] [pid 19] [client 172.19.0.1:48996] AH01114: HTTP: failed to make connection to backend: mock-auth, referer: http://localhost:8000/en/admin
172.19.0.1 - - [26/May/2022:14:19:30 +0000] "GET /oxauth/authorize?client_id=fake-development-client-key&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fauth-callback&response_type=code&scope=openid+offline_access&state=ikKZ0Gp4LesN2YJNcs8KejEEds56qbcDW5vpqP2W&nonce=PfMto2iKNoZlybwzFo1J0IJPC8uF1xp0GMF8xHhs&acr_values=gckey&ui_locales=en-CA+en HTTP/1.1" 503 566 "http://localhost:8000/en/admin" "Mozilla/5.0 (X11; Linux x86_64; rv:100.0) Gecko/20100101 Firefox/100.0"
```
I think it's a port mixup?
image

**Comment by tristan-orourke** [comment]
> Is the API server using the cookie api_session or XSRF-TOKEN to track the returning user one their way back from auth server?

The api_session is used only to choose where to redirect. If the api_session cookie disappeared between going to the auth server and coming back, authentication should still work fine.

I don't think we're using XSRF-TOKEN at all? Laravel may be setting it by default.

**Comment by tristan-orourke** [comment]
By the time I've found time to read through this, I think Peter has already asked all the really important questions! 

I'm really curious to understand the proxy better, I'll find some time to talk through a few questions about it.

**Comment by patcon** [comment]
Thanks @petertgiles! I'm spinning up fresh and investigating.

Musings: wondering if it has to do with us using `links` in docker-compose.yml rather than `networks` (which i think supercedes `links`). I had been toying with that during my trying to get it working, and maybe something persisted that made it work for me but not you 🤔 

Also, since `links` is deprecated, maybe our `docker-compose` version matters (or even if we're using `docker compose` (the docker-integrated command) vs `docker-compose` (standalone).

What version and commands are you using? I'm using `docker-compose`, and these versions:

```
patcon at Patricks-MacBook-Pro in ~/repos/gc-digital-talent (feature/2722-make-email-editable-and-nullable)
$ docker-compose --version
docker-compose version 1.29.2, build 5becea4c

patcon at Patricks-MacBook-Pro in ~/repos/gc-digital-talent (feature/2722-make-email-editable-and-nullable)
$ docker --version
Docker version 20.10.13, build a224086

patcon at Patricks-MacBook-Pro in ~/repos/gc-digital-talent (feature/2722-make-email-editable-and-nullable)
$ docker compose version
Docker Compose version v2.3.3
```

**Comment by patcon** [comment]
https://recordit.co/pCmAXSuZp2

Hm. Not sure what's up @petertgiles, I built fresh into /tmp, and just ran this, and it works as above:

```
cd /tmp
git clone GCTC-NTGC/gc-digital-talent
cd gc-digital-talent
git pr checkout 2813
docker-compose up --detach --build
docker-compose run --rm maintenance bash setup.sh
```

**Comment by tristan-orourke** [comment]
Just added "run `docker-compose run --rm maintenance bash refresh_api.sh` to install new composer dependencies" to Steps to Test

**Comment by tristan-orourke** [comment]
Hmm, it works fine fine for me, no problems

**Review by tristan-orourke — APPROVED** [review]

**Review by petertgiles — APPROVED** [review]
Hmm, I'm not sure what went wrong for me.  I tried again today (maybe more carefully?) and it ran fine!  I also deployed it to the dev server and it seems OK.  It's still up there now if you want to poke at it.

**Comment by petertgiles** [comment]
Here are my tips if you want to include them in your "help people to get their environment going again" post:
- delete the auth folder manually before checking out
- do `docker-compose down` before checking out
- add `--build` to your docker-compose up -d

**Comment by patcon** [comment]
k, thanks peter! Adapted from your recommendations (but favouring after-pull steps in case ppl don't notice msg)

(added `auth/` to gitignore, so auth won't risk being recommitted, and unmerged feature branches still work)

After pull:

1. `docker-compose down --remove-orphans`
2. `docker-compose up --detach --rebuild`

**Pull request merged**
